File: java/src/main/java/io/webweavex/fetch/Crawler.java
package io.webweavex.fetch;
import java.util.*;
import java.util.regex.*;
public class Crawler {
    private final HttpTransport transport;
    private final Set<String> visited = new LinkedHashSet<>();
    private final List<String> discovered = new ArrayList<>();
    public Crawler() { this.transport = HttpTransport.getDefault(); }
    public Crawler(HttpTransport t) { this.transport = t; }
    public Map<String, Object> crawl(String url) { crawlRecursive(url, 0); Map<String, Object> r = new LinkedHashMap<>(); r.put("visited", new ArrayList<>(visited)); r.put("discovered", new ArrayList<>(discovered)); return r; }
    private void crawlRecursive(String url, int depth) {
        if (depth > 3 || visited.contains(url)) return;
        visited.add(url);
        Map<String, Object> resp = transport.fetchSync(url);
        if (!(boolean) resp.getOrDefault("ok", false)) return;
        String text = (String) resp.get("text");
        if (text == null) return;
        Matcher m = Pattern.compile("href=\"([^\"]+)\"").matcher(text);
        while (m.find()) { String link = m.group(1); if (!visited.contains(link)) discovered.add(link); }
    }
}

File: kotlin/src/main/kotlin/io/webweavex/fetch/Crawler.kt
package io.webweavex.fetch

class Crawler(private val transport: HttpTransport = HttpTransport.default()) {
    private val visited = linkedSetOf<String>()
    private val discovered = mutableListOf<String>()

    suspend fun crawl(url: String, maxDepth: Int = 3): Map<String, Any> {
        crawlRecursive(url, 0, maxDepth)
        return mapOf("visited" to visited.toList(), "discovered" to discovered.toList())
    }

    private suspend fun crawlRecursive(url: String, depth: Int, maxDepth: Int) {
        if (depth > maxDepth || url in visited) return
        visited.add(url)
        val resp = transport.fetch(url)
        if (resp["ok"] != true) return
        val text = resp["text"] as? String ?: return
        Regex("href=\"([^\"]+)\"").findAll(text).forEach { m ->
            val link = m.groupValues[1]
            if (link !in visited) discovered.add(link)
        }
    }
}
